parlai/agents/starspace_torch/starspace_torch.py

```diff
@@ -0,0 +1,367 @@
+            margin=opt['margin'], size_average=False
+        )
+        self.reset()
+
+    def build_model(self):
+        print("[ creating StarspaceTorchAgent ]")
+        # this is not a shared instance of this class, so do full init
+        if (self.opt.get('model_file') and
+                (os.path.isfile(self.opt.get('model_file') + '.dict')
+                 or (self.opt['dict_file'] is None))):
+            # set default dict-file if not set
+            self.opt['dict_file'] = self.opt['model_file'] + '.dict'
+        # load dictionary and basic tokens & vectors
+        self.dict = DictionaryAgent(self.opt)
+
+        self.model = Starspace(self.opt, len(self.dict), self.dict)
+        if self.opt.get('model_file') and os.path.isfile(self.opt['model_file']):
+            self.load(self.opt['model_file'])
+            self.reset()
+        else:
+            self._init_embeddings()
+        self.model.share_memory()
+        if self.use_cuda:
+            self.model.cuda()
+
+    def train_step(self, batch):
+        """Train on a single batch of examples."""
+        if batch.text_vec is None:
+            return
+        batchsize = batch.text_vec.size(0)
+        self.model.train()
+        self.optimizer.zero_grad()
+
+        if not batch.candidates:
+            print('No negatives yet!')
+            return Output()
+
+        cands, cand_vecs, label_inds = self._build_candidates(
+            batch, source=self.opt['candidates'], mode='train')
+
+        if self.opt.get('input_dropout', 0) > 0:
+            batch_text, batch_label, batch_cand_vecs = self.input_dropout(
+                batch.text_vec,
+                batch.label_vec,
+                cand_vecs
+            )
+            xe, ye = self.model(
+                batch_text,
+                ys=batch_label,
+                cands=batch_cand_vecs
+            )
+        else:
+            xe, ye = self.model(
+                batch.text_vec,
+                ys=batch.label_vec,
+                cands=cand_vecs
+            )
+        xe = torch.cat([xe for _ in range(ye.size(1))], dim=1)
+
+        if self.use_cuda:
+            y = -torch.ones(xe.size(0), xe.size(1)).cuda()
+            y_0 = torch.ones(xe.size(0)).cuda()
+        else:
+            y = -torch.ones(xe.size(0), xe.size(1))
+            y_0 = torch.ones(xe.size(0))
+        y[:, 0] = y_0
+
+        loss = self.criterion(
+            xe.view(-1, xe.size(-1)),
+            ye.view(-1, xe.size(-1)),
+            y.view(-1)
+        )
+        loss.backward()
+        self.optimizer.step()
+        scores = nn.CosineSimilarity(dim=-1).forward(xe, ye)
+
+        # Update metrics
+        self.metrics['loss'] += loss.item()
+        self.metrics['examples'] += batchsize
+        _, ranks = scores.sort(1, descending=True)
+        for b in range(batchsize):
+            rank = (ranks[b] == label_inds[b]).nonzero().item()
+            self.metrics['rank'] += 1 + rank
+
+        return Output()
```

Comment:
Stuff needed here I believe (you compute the ranks, so might as well return something)

Reply:
doesn't really make sense to returning anything here, as during training we are ranking a set of custom candidates and not the candidates provided by the task

Reply:
(y)